Customizer: Improve menu creation flow

brandonpayton opened the pull request:
# Overview

This is a personal PR to track progress on [trac #40104](https://core.trac.wordpress.org/ticket/40104).

# Development

Here's a rough breakdown to help me focus and track progress.

* [x] Update Menus layout
* [x] Add New Menu section
* [x] Update Edit Menu layout and text
* [x] Update Locations layout
* [ ] Clean up styles -- separating this because it is more efficient to consider styles all at once
* [x] Find out how to create localized strings that contain singular or plural nouns depending on the data
* [ ] Unit tests
* [ ] Review localized string calls for clarity and correctness
* [ ] Review accessibility

# Latest sketch

![customizer-menu-flow-v2](https://user-images.githubusercontent.com/530877/30777303-762c68c2-a06c-11e7-8f97-12ed4779b499.png)

# Questions for design

* [ ] The locations view has a "If your theme supports widgets" area that links to the widget menu and currently only shows when a theme supports widgets. Should we unconditionally show this paragraph or remove the "if" since it will only be shown when widgets are supported?
    * **NOTE**: If we unconditionally show, we'll probably need to make the link conditional upon widget support. Otherwise we'll link to an unsupported customization view.

# Development follow-up

These are improvements that need to be made after initial development. I'm adding them here so I can avoid being distracted while doing the initial work and learning about WordPress.

* [ ] Should the locations section determine all of its description based on `$num_locations`? Right now, it is split across two files.
* [ ] Make sure "Menu Location" and "Menu Locations" labels are correctly applied depending for singular and plural locations. 
* [ ] Improve names of different location-related sections and controls to clearly differentiate.
* [ ] Rename `WP_Customize_New_Menu-*` types to use `WP_Customize_Nav_Menu-` prefix instead so sections and controls are grouped with others pertaining to nav menus.
* [ ] Restore double-border that used to appear at the top of the All Locations section title and the first Menu section title.
* [ ] Explore creating a generic button component with a description
* [ ] Add documentation with `@since` tag to all new JS declarations
* [ ] Does WP_Customize_Nav_Menu_Control actually need to check whether a theme supports menus? Shouldn't that check be done by a parent?
* [ ] Focus name field when expanding new menu section
* [ ] DRY use of "Custom Menu widget" in the locations panel to make sure we show the same translated name here and in the widgets panel.

brandonpayton commented:
I mentioned this in the WordPress.org Slack in the `core-customize` channel. It would be good to get feedback on overall structure, but please note that this is a work in progress.

westonruter commented on src/wp-includes/class-wp-customize-control.php:
```diff
@@ -739,14 +739,24 @@ protected function content_template() {}
 
 /**
- * WP_Customize_New_Menu_Control class.
+ * WP_Customize_Nav_Menu_Auto_Add_Control class.
```
Class name here is incorrect.

brandonpayton replied:
Ooh, embarrassing. Fixed.

westonruter commented on src/wp-admin/js/customize-nav-menus.js:
```diff
@@ -2788,8 +2895,10 @@
 		nav_menu_name: api.Menus.MenuNameControl,
+		nav_menu_locations: api.Menus.MenuLocationsControl,
 		nav_menu_auto_add: api.Menus.MenuAutoAddControl,
-		new_menu: api.Menus.NewMenuControl
```
Deletion of the `NewMenuControl` (in JS and PHP) _may_ have back-compat problems, but I hope no plugins have tried interacting with these directly. Probably ok to delete.

brandonpayton replied:
Yeah, I wondered about this as well.

westonruter commented on src/wp-admin/js/customize-nav-menus.js:
```diff
@@ -2121,6 +2332,31 @@
+	 * wp.customize.Menus.MenuDeleteControl
+	 *
+	 * Customizer control for deleting a nav menu.
+	 *
```
Add `@since` for class and methods.

brandonpayton replied:
When creating new overrides for inherited methods, is it better to use `@since` or `@inheritsdoc`. I'm seeing a mix of both and neither in the Customizer JS.

westonruter replied:
Probably _both_ should be included actually. The override gets introduced in a version so it should get `@since` but then there shouldn't be a need to duplicate all of the other JSDoc, and in that case `@inheritsdoc` should be included as well, but with additional description for what the override is doing.

brandonpayton replied:
I was just thinking about that and agree with your reasoning. Unfortunately, it looks like `@inheritsdoc` is all or nothing. From http://usejsdoc.org/tags-inheritdoc.html:
*The @inheritdoc tag indicates that a symbol should inherit its documentation from its parent class. Any other tags that you include in the JSDoc comment will be ignored.*

But maybe the above behavior is acceptable for generated documentation, and we can still include `@since` for people reading the source.

brandonpayton replied:
Since we needed to make a choice, I went ahead and used @since instead of @inheritsdoc because it is the prevailing practice in `customize-nav-menus.js`.

If we used the `@inheritsdoc` with `@since` and a description, the latter two would be lost when actually generating documentation. That said, I'm not sure whether we actually generate docs based on our JSDoc.

If you'd like me to take a different approach, please let me know.

westonruter commented on src/wp-admin/js/customize-nav-menus.js:
```diff
@@ -2027,45 +2187,96 @@
+				* preview's sending of the activeControls to result in this control
+				* being deactivated.
+				*/
+				control.active.validate = function() {
```
This should be able to be eliminated entirely now because of this change which landed in 4.7: https://github.com/WordPress/wordpress-develop/blob/4666f2fd393a3803fadae4a5aea3cf6e32973d9b/src/wp-admin/js/customize-controls.js#L5842-L5854

brandonpayton replied:
Thanks! Done.

westonruter commented:
@celloexpressions This would benefit a lot from your review, given your work on Menu Customizer.

celloexpressions commented on src/wp-admin/css/customize-controls.css:
```diff
@@ -333,6 +333,15 @@ body.outer-section-open .wp-full-overlay.expanded .wp-full-overlay-main {
 	top: 46px;
 }
 
+#customize-controls .customize-section-title-heading {
```
We should generally avoid using generic selectors like this when changing specific features, and instead scope to a menus-specific selector.

brandonpayton replied:
Great point. I'll fix it. Thanks!

brandonpayton replied:
I think my original thought was that this might be a more generic pattern, but I don't think it is yet.

celloexpressions commented on src/wp-admin/css/customize-controls.css:
```diff
@@ -333,6 +333,15 @@ body.outer-section-open .wp-full-overlay.expanded .wp-full-overlay-main {
 }
 
+#customize-controls .customize-section-title-heading {
+	margin: 15px;
```
The section typically handles left/right margins with 12px padding (not within a control), and top/bottom margins are typically 15px.

brandonpayton replied:
This seems to be a novel case since these are children of `.control-panel-content` rather than `.accordion-section-content` which applies the `12px` padding. It's very possible I'm missing something, but it seems like the padding needs to be on these child elements because otherwise we pad the section title.

<img width="303" alt="screen shot 2017-10-03 at 11 31 47 am" src="https://user-images.githubusercontent.com/530877/31142218-794b9dce-a82e-11e7-898d-b12ff52d3b97.png">

celloexpressions commented on src/wp-admin/css/customize-nav-menus.css:
```diff
@@ -738,27 +723,16 @@ body.adding-menu-items #customize-preview iframe {
-#accordion-section-add_menu .accordion-section-title {
-	padding-left: 45px;
+#accordion-section-add_menu h3,
+#accordion-section-add_menu button.customize-add-menu-button {
```
No need to target the `button` element in the selector, especially when there's `button` in the class name as well. If specificity is needed, consider adding an additional container.

brandonpayton replied:
I tend to avoid element selectors and thought I followed a pattern I'd observed elsewhere in the CSS. There are some instances but not many. In any case, I'll fix it.

celloexpressions commented on src/wp-admin/css/customize-nav-menus.css:
```diff
@@ -738,27 +723,16 @@ body.adding-menu-items #customize-preview iframe {
-	position: absolute;
-	top: 12px;
-	left: 14px;
-	content: "\f132";
```
Related to the design, I'm not excited about losing visual parity with "add menu items" and "add widgets" by removing the icon here.

celloexpressions commented on src/wp-admin/js/customize-nav-menus.js:
```diff
@@ -950,7 +950,7 @@
 
 		populateControls: function() {
-			var section = this, menuNameControlId, menuAutoAddControlId, menuControl, menuNameControl, menuAutoAddControl;
+			var section = this, menuNameControlId, menuLocationsControlId, menuAutoAddControlId, menuDeleteControlId, menuControl, menuNameControl, menuLocationsControl, menuAutoAddControl, menuDeleteControl;
```
I'm skeptical that we need to add this many controls for each menu. Without digging in in detail, my initial though was that we shouldn't need significant changes to tweak the UI as the ticket suggested. If we do need all of these, it's probably time to break this into multiple lines.

brandonpayton replied:
I also meant to say that I'd appreciate any ideas you have about how to simplify.

celloexpressions commented on src/wp-admin/js/customize-nav-menus.js:
```diff
@@ -1100,58 +1142,172 @@
 	 * @augments wp.customize.Section
 	 */
 	api.Menus.NewMenuSection = api.Section.extend({
-
```
I think we need this line break per the JS coding standards; should avoid changes like this unless they're needed or improve readability, generally speaking.

brandonpayton replied:
Understood. The change might have been almost unconscious, but I shouldn't have committed it.

The line is restored.

celloexpressions commented on src/wp-admin/js/customize-nav-menus.js:
```diff
@@ -1481,7 +1637,9 @@
 
-			$removeBtn.on( 'click', function() {
+			$removeBtn.on( 'click', function(event) {
+				event.preventDefault();
```
Make sure this is a proper button with a type attribute - preventDefault() shouldn't be needed here now at a glance.

brandonpayton replied:
You're right. Thanks for catching this.

celloexpressions commented on src/wp-admin/js/customize-nav-menus.js:
```diff
@@ -2679,105 +2863,25 @@
 
 	/**
-	 * wp.customize.Menus.NewMenuControl
+	 * wp.customize.Menus.NewMenuSubmitControl
```
Do we need to rename this? Or, remove/rework it to this extent? If there's a way to minimize the code/API changes, that would be preferable, although this may be the best way to do it.

brandonpayton replied:
It wasn't intended as a rename. It just diffs that way. The `NewMenuControl` was the text input and button combination that expanded under the old "Add Menu" button. I created `NewMenuSubmitControl` for a description + button combination because I didn't know of a cleaner way to do it with the current infrastructure.

As you mentioned elsewhere, a generic button component with a description would be better.  I'm going to continue roughing out the current design, at least for a concrete reference point for conversation, but I am adding a reminder to look at creating the more generic component after that.

celloexpressions commented on src/wp-includes/class-wp-customize-nav-menus.php:
```diff
@@ -400,34 +400,35 @@ public function enqueue_scripts() {
+				/* translators: %s:      menu location */
+				'menuLocation'           => _x( '(Currently set to: %s)', 'menu' ),
+				'menuNameLabel'          => __( 'Menu Name' ),
+				'newMenuNameDescription' => __( 'If your theme has multiple menus, giving them clear names will help you manage them.' ),
```
Can this get a shorter name to avoid changing so many lines? Not critical and better to keep it aligned, but history is harder to trace and things are more likely to get broken when more things change (in this case, spacing for the entire block of labels).

brandonpayton replied:
Yeah, I didn't like that either, but I haven't yet thought of a shorter name that doesn't sacrifice clarity. Any ideas?

celloexpressions commented on src/wp-includes/class-wp-customize-nav-menus.php:
```diff
@@ -533,12 +534,19 @@ public function customize_register() {
 		}
 
+		// Require JS-rendered section types.
+		$this->manager->register_section_type( 'WP_Customize_Nav_Menu_Locations_Section' );
```
These should go directly under `register_panel_type` below, before the list of `register_control_type`s. And the comment there can say panel, section, and control types.

brandonpayton replied:
I missed that the below section wasn't just control registration. Fixed.

celloexpressions commented on src/wp-includes/class-wp-customize-nav-menus.php:
```diff
@@ -560,22 +568,24 @@ public function customize_register() {
 		$num_locations = count( array_keys( $locations ) );
 		if ( 1 == $num_locations ) {
-			$description = '<p>' . __( 'Your theme supports one menu. Select which menu you would like to use.' ) . '</p>';
+			$description = '<p>' . __( 'Your theme can display menus in one location. Select which menu you would like to use.' ) . '</p>';
```
We may want to break out the text changes since they could be made for 4.9 but I'm not sure about the rest being ready.

celloexpressions commented on src/wp-includes/class-wp-customize-section.php:
```diff
@@ -375,6 +372,23 @@ protected function render_template() {
+			<span class="screen-reader-text"><?php _e( 'Press return or enter to open this section' ); ?></span>
+		</h3>
+		<?php
+	}
```
This general API change should have a distinct ticket and patch for discussion.

brandonpayton replied:
Yeah, this is the most concerning part of this PR to me for 4.9. I'll see if I can find a reasonable way to live without the change.

celloexpressions commented on src/wp-includes/class-wp-customize-section.php:
```diff
@@ -383,6 +397,9 @@ protected function render_template() {
 require_once( ABSPATH . WPINC . '/customize/class-wp-customize-sidebar-section.php' );
 
+/** WP_Customize_Nav_Menu_Locations_Section class */
+require_once( ABSPATH . WPINC . '/customize/class-wp-customize-nav-menu-locations-section.php' );
```
My initial gut reaction is that we shouldn't need a custom section class for the menu locations section. Haven't looked at what it does yet.

brandonpayton replied:
This is the section for View All Locations, and it is custom because there is a custom heading and description preceding the usual section title. Please let me know if there is a better way to do this.

celloexpressions commented on src/wp-includes/customize/class-wp-customize-nav-menu-control.php:
```diff
@@ -44,6 +36,7 @@ public function render_content() {}
 	 */
 	public function content_template() {
 		?>
+		<p class="new-menu-item-invitation"><?php _e( 'Time to add some links! Click "Add menu items" to start putting pages, categories, and custom links in your menu. Add as many things as you\'d like.' ); ?></p>
```
Another potential thing to do as a quick change for 4.9 if the rest isn't ready.

celloexpressions commented on src/wp-includes/customize/class-wp-customize-nav-menu-delete-control.php:
```diff
@@ -0,0 +1,48 @@
+ *
+ * @see WP_Customize_Control
+ */
+class WP_Customize_Nav_Menu_Delete_Control extends WP_Customize_Control {
```
This should be explored as a more generic "button" control unless there's a large amount of JS that makes more sense as part of this control than as part of the section (I'd lean toward managing this in the section JS, with a control for the visible UI only, since it deletes the entire section). I wonder if you can get a good button out of the base customize control if your `type` is "button" (with `input_attrs` for classes, etc.)?

brandonpayton replied:
I looked at the possibility of using the base customize control but was concerned the the description I wanted would appear after the control rather than before, as the design requires. I agree this seems like it should be unnecessary, but it was the best solution I found to match the design.

westonruter replied:
@brandonpayton  The template here can now be moved to a `WP_Customize_Nav_Menus::print_templates()` method and then referenced in the new `templateId` param to avoid having to create an entirely new PHP class file. An example of this can be seen with the preview link control and the trash control, the latter which is a button itself.

celloexpressions commented on src/wp-includes/customize/class-wp-customize-nav-menu-locations-control.php:
```diff
@@ -0,0 +1,61 @@
+ *
+ * @since 4.9.0
+ */
+class WP_Customize_Nav_Menu_Locations_Control extends WP_Customize_Control {
```
This makes more sense as its own control type than some of the other things.

celloexpressions replied:
We should document changes like this that are refactoring-oriented carefully, and emphasize that there are underlying technical changes like this in addition to the UI changes on trac, make/core, etc.

brandonpayton replied:
That's a good point. I would prefer to deliver changes like this refactoring and the addition of `title_content()` to `WP_Customize_Section` separately. I suppose it depends on how important we feel this enhancement is at this point.

celloexpressions commented on src/wp-includes/customize/class-wp-customize-nav-menu-locations-section.php:
```diff
@@ -0,0 +1,70 @@
+ *
+ * @see WP_Customize_Section
+ */
+class WP_Customize_Nav_Menu_Locations_Section extends WP_Customize_Section {
```
It feels like this one shouldn't need to be a custom section type. Can we get it to work with some additional menus-panel-managed JS (since menu locations also interact with the other menu sections in that panel), and creative manipulation of the standard title and description properties?

brandonpayton replied:
I will take a look at this shortly. I'd love to do without the additional type or the `title_content()` change to `WP_Customize_Section`.

celloexpressions commented on src/wp-includes/customize/class-wp-customize-new-menu-control.php:
```diff
@@ -1,38 +0,0 @@
- *
- * @see WP_Customize_Control
- */
-class WP_Customize_New_Menu_Control extends WP_Customize_Control {
```
This is an example of a control type that probably didn't need to exist in the first place. I think it was in 4.3 initially despite the conflicting `@since`.

celloexpressions commented on src/wp-includes/customize/class-wp-customize-new-menu-submit-control.php:
```diff
@@ -0,0 +1,45 @@
+ *
+ * @see WP_Customize_Control
+ */
+class WP_Customize_New_Menu_Submit_Control extends WP_Customize_Control {
```
This also feels redundant to a more generic button control with a title and description parameter.

brandonpayton replied:
Agreed. I'd prefer a generic button control with a description that precedes the button. I thought that where the description should appear might vary on a case-to-case basis, but after further thought, I think it may be most natural for descriptions to precede buttons.

I'd like to look at this, but I'm wondering whether we can tolerate the presence of a specialized control such as this in exchange for the enhancement to 4.9. We could create an issue, leave a `@todo`, or do both to that effect.

westonruter replied:
@brandonpayton Per above, a generic button control can now be easily added with a template. See changeset trash control.

brandonpayton commented:
I updated the description to reduce scope prior to the 4.9 beta deadline.

brandonpayton commented:
I updated the styles locally and am happy with the appearance, but I need to be more careful to use existing classes appropriately rather than creating one-off rules. I'm planning to take a fresh look at this in the morning.

brandonpayton commented:
I went ahead and pushed existing style progress and a fix for a unit test.

All the tests are passing, but I've noticed at least one that is broken. It is checking for control order on the panel, and I think it is actually verifying the order the controls were added on the server-side, instead of verifying the actual order in the UI. It's asserting that the locations control is first in the menu panel and passing, but when rendered, the locations control is last in the panel due to priority.

brandonpayton commented:
I rebased this PR on the latest from trunk, including @westonruter's improvements to the Customizer JS API. I still have a good amount of cleanup to do, but if anyone is interested in testing this, I think it's ready.

I'd like to post a patch but should wait for morning. It's late, and I'm going to turn in.

Thanks for your thought and attention to this.

brandonpayton commented:
I reviewed the updates and did some additional cleanup.

The last big thing I am aware of is to revert the addition of `title_template()` to `WP_Customize_Section` and find a less invasive way to accomplish the same thing for 4.9.

westonruter commented on src/wp-admin/js/customize-nav-menus.js:
```diff
@@ -2210,6 +2444,11 @@
 			}
+
+			// Wait until we know existing menu items have been added
+			api.bind( 'ready', function () {
```
I don't think this is right. The `MenuControl` may be added before `ready` and after `ready`, so then there will be event handlers for `ready` that may never get triggered.

westonruter replied:
So I think instead what you may want to do is a `_.defer()` because what you really want is a nextTick for all of the items to be added.

brandonpayton replied:
I was hoping for a way to explicitly say "after everything is there", but since it sounds like `_.defer()` is tolerable, I'll switch to that for 4.9.

westonruter commented on src/wp-includes/customize/class-wp-customize-nav-menu-locations-section.php:
```diff
@@ -0,0 +1,70 @@
+	protected function title_template() {
+		?>
+		<span class="customize-control-title customize-section-title-menu_locations-heading"><?php _e( 'Menu Locations' ); ?></span>
+		<p class="customize-control-description customize-section-title-menu_locations-description">{{ data.titleDescription }}</p>
```
One option instead of introducing `title_template` here would be to inject these elements with JS in the section's JS `Section` subclass.

westonruter commented on src/wp-includes/customize/class-wp-customize-nav-menus-panel.php:
```diff
@@ -92,6 +92,10 @@ protected function content_template() {
+		<?php
+		// NOTE: The following is a workaround for an inability to treat (and thus label) a list of sections as a whole.
+		?>
+		<li class="customize-control-title customize-section-title-nav_menus-heading">Menus</li>
```
This needs to be translated.

westonruter commented on src/wp-includes/customize/class-wp-customize-new-menu-submit-control.php:
```diff
@@ -0,0 +1,45 @@
+	protected function content_template() {
+		?>
+		<p id="customize-new-menu-submit-description"><?php _e( 'Click "next" to start adding links to your new menu.' ); ?></p>
+		<button id="customize-new-menu-submit" type="button" class="button" aria-describedby="customize-new-menu-submit-description"><?php _e( 'Next' ); ?></button>
```
Per above, I think the button could be just injected into the template via JS instead of creating a new PHP subclass.

brandonpayton commented:
With @westonruter's help, I removed the specialized control and section types that are no longer necessary and reverted addition of the `title_template()` method to `WP_Customize_Section`.

westonruter commented:
Committed to `trunk` in r41768.
